### ISSUE TITLE
Show image uris as actual images inline in messages

### ISSUE THREAD

**Issue by nikolaiwarner** [cabal-club/cabal-desktop#12]
If we can detect that a uri sent within a message is an image, we should try to show that image inline in the message.

And also have an option to enable/disable the feature.

**Comment by masukomi** [comment]
given the emphasis on encrypted chats, calling out to an external url to download an image is a potential security risk for people discussing sensitive material.  It would expose their IPs and other potential metadata

This could work if there was some "click to load image" type thing. 

I think a better solution would be to actually transmit the images via cabal. I don't know much about dat, but It would be good if it could be done similarly to how SSB does it where we can optionally download the blobs. that way folks on low bandwidth connections could still participate in the chat without getting bogged down trying to download an image.

**Comment by noffle** [comment]
I'd really like it if we had a good p2p blob resolver, so you could post all sorts of things:

- `dat://...`
- `ipfs://...`
- `ssb://...`
- `magnet://...`

and then it would check if you have/use that p2p network and download it when you'd like it.

If we really want to keep data within the network, we could use a hyperdb trick where each user has a separate "content feed" that they can post binary blobs to. Unlike the main message feed (which is what we use to send chat messages), the content feed would be opt-in: you'd need to explicitly ask for entries in it. This functionality comes for free with [hypercore](https://github.com/mafintosh/hypercore), which cabal is built on.

**Comment by cblgh** [comment]
+1 i think starting out with the content feed for cabal makes sense. 

after we have something working we could work on the more general p2p blob resolver module (which would be super rad for many projects in our neck of the woods)

**Comment by todrobbins** [comment]
#14 addresses this feature request by creating a dat archive for shared files. I agree that have a toggle to hide the images would be necessary, especially to hide flashy GIFs, etc.

**Comment by warcode** [comment]
As someone who only cares about the confidentiality of message content and not metadata, and only would hang out with friends, _optional_ direct content embeds seems like it would be a good idea? At least until whatever perfect solution is planned can be implemented.

**Comment by cblgh** [comment]
You’re right, that could be something people could toggle on in the settings.

**Comment by cblgh** [comment]
@khubo raises the **excellent** idea of using @rangermauve's [p2p-fetch](https://github.com/RangerMauve/p2p-fetch) for allowing cross-protocol fetching of images 🧠🤯

**Comment by RangerMauve** [comment]
If you use electron you can integrate it into the protocol handlers so you could just put urls into the `src` attribute. :o

**Comment by cblgh** [comment]
@RangerMauve we do use electron! have you done that in agregore somewhere?

**Comment by noffle** [comment]
great idea @RangerMauve on p2p-fetch! it'd be great to build this out to support fetching and publishing of {dat,ipfs,hypercore,etc}

**Comment by RangerMauve** [comment]
I have some code for converting fetch interfaces to protocol handlers for Electron here: https://github.com/AgregoreWeb/agregore-browser/blob/master/app/protocols/fetch-to-handler.js

I combine that with `dat-fetch` here: https://github.com/AgregoreWeb/agregore-browser/blob/master/app/protocols/hyper-protocol.js

And with `ipfs-fetch` here: https://github.com/AgregoreWeb/agregore-browser/blob/master/app/protocols/ipfs-protocol.js

I register the handlers on the electron session here: https://github.com/AgregoreWeb/agregore-browser/blob/master/app/protocols/index.js#L72

With those in place you can use the `fetch()` API in the electron side with the protocols without needing to bundle extra libraries, and any image or video src tags will automatically work with `hyper://` and `ipfs://` URLs.

Yeah, I'd like to work on p2p-fetch a bit more soon 😁

There's a bug in Electron 11 where the response headers aren't properly available for custom protocols, but that might still work in electron 9 or whatever this is using.
